keps/sig-storage/2268-non-graceful-shutdown/README.md

```diff
@@ -0,0 +1,445 @@
+
+## Summary
+
+In this KEP, we are proposing a way that allows stateful workloads to failover to a different node successfully after the original node is shutdown or in a non-recoverable state such as the hardware failure or broken OS.
+
+## Motivation
+
+The Graceful Node Shutdown [KEP](https://github.com/kubernetes/enhancements/tree/master/keps/sig-node/2000-graceful-node-shutdown) introduced a way to detect a node shutdown and handle it gracefully. However, a node shutdown action may not be detected by Kubelet's Node Shutdown Mananger, either because the command does not trigger the inhibitor locks mechanism used by Kubelet or because of a user error, i.e., the ShutdownGracePeriod and ShutdownGracePeriodCriticalPods are not configured properly.
+
+In this KEP, we are proposing a way to handle node shutdown cases that are not detected by Kubelet's Node Shutdown Manager.
+
+### Use Cases
+
+* User wants a stateful workload to failover to another running node in the following cases with minimum intervention when it's safe to do so.
+  * The node is shutdown
+  * The node is in other non-recoverable cases:
+    * hardware failure
+    * broken OS
+
+### Goals
+
+* Increase the availability of stateful workloads when a node is shutdown, or in other non-recoverable cases such as the hardware failure or broken OS.
+* Automate self-healing for stateful workloads.
+
+### Non-Goals
+
+* Node/control plane partitioning other than a node shutdown is not covered by the proposal, but will be addressed in the future and built on top of this design.
+  * We do not have a way to determine whether it is safe to detach in the case of node partitioning. In the Alternatives section, we discussed the node fencing approach and explained why that is not selected for this KEP but will be considered after this KEP is implemented.
+* Implement in-cluster logic to handle node/control plane partitioning.
+
+## Proposal
+
+In this section, we are describing the proposed changes to enable workloads to failover to a running node when the original node is shutdown but not detected by Kubelet's Node Shutdown Manager, or when the node is in other non-recoverable cases such as the hardware failure or broken OS.
+
+Existing logic:
+1. When a node is not reachable from the control plane, the health check in Node lifecycle controller, part of kube-controller-manager, sets Node v1.NodeReady Condition to False or Unknown (unreachable) if lease is not renewed for a specific grace period. Node Status becomes NotReady.
+
+1. After 300 seconds (default), the Taint Manager tries to delete Pods on the Node after detecting that the Node is NotReady. The Pods will be stuck in terminating status.
+
+Proposed logic change:
+1. [Proposed change] This proposal requires a user to apply a `quarantine` taint on a node when the user has confirmed that this node is shutdown or in a non-recoerable state due to the hardware failure or broken OS.
```

Comment:
This says "user" - does it mean "or controller" ?

Reply:
Why is it a taint?  Taints imply that a toleration can overcome it, but if it is off...

Reply:
Reading this it feels more like a `node.status.powerState` field than a taint.  E.g. `powerState = ConfirmedOff".

The scheduler would not select nodes that are not Ready, which would necessarily include powered-off.

Cloud controllers would be responsible for poking at NotReady nodes and seeing if they can be confirmed as off. (and a logical next step would be to turn them off after X time, when possible -> fencing).

Reply:
In the previous design, we are thinking to allow either user manually check machine is off or cloud provider automatically check machine status is shutdown. But there is a concern of checking automatically by cloud provider, whether it can differentiate between node is shutdown or it is in the middle of rebooting. So in this current design, we want to only allow user to manually check the status and make sure machine is off before adding the taint.

Yes, the taint here is mainly a signal about the situation so that controller can take actions based on this signal. It is not necessary to use taint, but it is easy to use (allowing either user or controller could update it) in this situation. It is a good point that in case there is some tolerations setting which might overwrite the logic here. Annotation is another option here, but it seems not recommended due to its limitations?

In this KEP we want to reduce the scope so that we can make some progress for this alpha feature. There are some other user cases that we plan to address in the future, that could build upon on this, e.g., fencing, putting machine in maintenance mode explicitly for hardware upgrade etc.

Reply:
In this proposal, we propose to use "NoExecute" effect for quarantine taint of shutdown node. so I think there are two options about pod tolerations.
1. follow taint and toleration policy: If pod is set to tolerate all taints and effects, that means user does NOT want to evict pods when node is not ready. So GC controller will filter out those pods and only force delete pods that do not have a matching toleration.
2. ignore toleration setting: GC controller delete ALL pods that are scheduled on the shutdown node since there is point of letting pods stay in the node that is already shutdown. But this will break the taint and toleration policy.
So I think option 1 is preferable considering the taint and toleration policy.

Reply:
In this proposal, we propose to use "NoExecute" effect for quarantine taint of shutdown node. so I think there are two options about pod tolerations.
1. follow taint and toleration policy: If pod is set to tolerate all taints and effects, that means user does NOT want to evict pods when node is not ready. So GC controller will filter out those pods and only force delete pods that do not have a matching toleration.
2. ignore toleration setting: GC controller delete ALL pods that are scheduled on the shutdown node since there is point of letting pods stay in the node that is already shutdown. But this will break the taint and toleration policy.
So I think option 1 is preferable considering the taint and toleration policy.

Reply:
Updated KEP with option 1.